src/d3a/models/strategy/storage.py

```diff
@@ -381,28 +382,29 @@ def buy_energy(self, market):
             if offer.seller == self.owner.name:
                 # Don't buy our own offer
                 continue
+            # Check if storage has free capacity and if the price is cheap enough
+            if self.state.free_storage(market.time_slot) <= 0.0 \
+                    or (offer.price / offer.energy) > max_affordable_offer_rate:
+                continue
```

Comment:
In order to properly refactor this, I would suggest to split these 2 if clauses: 
- if there is no free storage, return (also consider to add this check at the very beginning of the buy_energy function)
- if the offer rate is greater than the max_affordable_offer_rate, continue

Reply:
You are totally right! Changing it.

Reply:
Yes, but `buy_energy` method handles a specific market (it receives it as an input argument). Therefore future markets will still be evaluated.